upgradeccl: deflake and unskip `TestTenantUpgradeFailure`

healthy-pod opened the pull request:
This code changes deflakes `TestTenantUpgradeFailure` by
expanding the list of expected errors to any non-nil error
when the tenant is is stopped in the middle of an upgrade.
It also unskips the test.

Closes https://github.com/cockroachdb/cockroach/issues/106279
Release note: None
Epic: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108388)
<!-- Reviewable:end -->

healthy-pod commented:
Stressed locally for 1000 runs / more than 1h 30m

knz commented on pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go:
```diff
@@ -278,6 +277,8 @@ func TestTenantUpgradeFailure(t *testing.T) {
 									ctx context.Context, version clusterversion.ClusterVersion, deps upgrade.TenantDeps,
 								) error {
 									tenantStopperChannel <- struct{}{}
+									// Wait for a second here to give the tenant some time to be stopped.
```
You could define a channel that gets `close()`d here, and wait on that below.

healthy-pod replied:
We already wait on `waitForTenantClose` which only receives a value once the stopper returns (i.e. once all workers including the `v2` migration return).

healthy-pod commented:
My initial thought was: we need to give the tenant more time to be stopped but then I learnt that the tenant stopper will wait for all workers to return before stopping which means it will wait until `v2` returns.

The stopper can't return if the v2 migration didn't return. Somehow however, just waiting in the v2 migration (using time.Sleep) after invoking the stopper helps and lets us get the expected error message. I can't explain why. We have two options: either the current change OR instead of expecting a specific error just expect that we get a non-nil error on the first upgrade attempt.

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @healthy-pod)

___
*[`pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go` line 284 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108388#-NbV3Wxl4QQmOb4yEUOb:-NbV3WxmEFDNuiFRHQ1S:b-jzhdtg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5dc225fcdd9b44decd6671b88e6474a604fed93/pkg/ccl/kvccl/kvtenantccl/upgradeccl/tenant_upgrade_test.go#L284)):*
> ```Go
> 										// Somehow waiting here helps and blocking the stopper for sometime helps
> 										// us get the right error message below [investigating why].
> 										time.Sleep(time.Second * 10)
> ```

Maybe if you need to keep this sleep you could combine it with a timer channel (`time.After()`) and a select on another channel that signals the tests has completed. This ensures we don't pay this time overhead when the test can complete earlier.


<!-- Sent from Reviewable.io -->

knz commented:
@healthy-pod could you review #108573. Then I would invite you to rebase this PR on top of #108573 to continue the investigation. The code will be easier to review that way.

Once that is done, I invite you to consider this patch: https://github.com/knz/cockroach/commit/8f85b23ed9b1acd4a42341aba6587182837b6c1c

I believe the use of `OnQuiesce` in that code will suitably replace the Sleep you had added earlier.

Commit:
upgradeccl: deflake and unskip `TestTenantUpgradeFailure`

This code changes deflakes `TestTenantUpgradeFailure` by
expanding the list of expected errors to any non-nil error
when the tenant is is stopped in the middle of an upgrade.
It also unskips the test.

Closes #106279
Release note: None
Epic: none

knz reviewed (approved):
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108388)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

healthy-pod commented:
TFTR!

bors r=knz
